onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+    const std::vector<std::string>& nodes_modes,
+    std::vector<int64_t> nodes_nodeids,
+    std::vector<int64_t> nodes_treeids,
+    std::vector<int64_t> nodes_truenodeids,
+    std::vector<OTYPE> nodes_values,
+    const std::string& post_transform,
+    std::vector<int64_t> class_ids,
+    std::vector<int64_t> class_nodeids,
+    std::vector<int64_t> class_treeids,
+    std::vector<OTYPE> class_weights,
+    std::vector<std::string> classlabels_strings,
+    std::vector<int64_t> classlabels_int64s) {
+  classlabels_strings_ = classlabels_strings;
+  classlabels_int64s_ = classlabels_int64s;
+  class_count_ = classlabels_strings_.size() == 0 ? classlabels_int64s_.size() : classlabels_strings_.size();
+  if (class_count_ == 0) {
+    ORT_THROW("classlabels_int64s_ and classlabels_strings_ cannot be both empty.");
+  }
+
+  TreeEnsembleCommon<ITYPE, OTYPE>::init(
+      aggregate_function, base_values, class_count_, nodes_falsenodeids,
+      nodes_featureids, nodes_hitrates, nodes_missing_value_tracks_true,
+      nodes_modes, nodes_nodeids, nodes_treeids, nodes_truenodeids,
+      nodes_values, post_transform, class_ids, class_nodeids,
+      class_treeids, class_weights);
+
+  std::set<int64_t> weights_classes;
+  weights_are_all_positive_ = true;
+  for (size_t i = 0, end = class_ids.size(); i < end; ++i) {
+    weights_classes.insert(class_ids[i]);
+    if (class_weights[i] < 0)
+      weights_are_all_positive_ = false;
+  }
+  binary_case_ = class_count_ == 2 && weights_classes.size() == 1;
+  if (classlabels_strings_.size() > 0) {
+    class_labels_.resize(classlabels_strings_.size());
+    for (size_t i = 0; i < classlabels_strings_.size(); ++i)
+      class_labels_[i] = i;
+  }
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommonClassifier<ITYPE, OTYPE>::compute(const Tensor* X, Tensor* Z, Tensor* label) const {
+  if (classlabels_strings_.size() == 0) {
+    this->compute_agg(
+        X, Z, label,
+        _AggregatorClassifier<ITYPE, OTYPE>(
+            this->roots_.size(), class_count_,
+            this->post_transform_, &(this->base_values_),
+            &(classlabels_int64s_), binary_case_,
+            weights_are_all_positive_));
+  } else {
+    int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
+    std::shared_ptr<IAllocator> allocator = std::make_shared<CPUAllocator>();
+    Tensor label_int64(DataTypeImpl::GetType<int64_t>(), TensorShape({N}), allocator);
+    this->compute_agg(
+        X, Z, &label_int64,
+        _AggregatorClassifier<ITYPE, OTYPE>(
+            this->roots_.size(), class_count_,
+            this->post_transform_, &(this->base_values_),
+            &(class_labels_), binary_case_,
+            weights_are_all_positive_));
+    const int64_t* plabel = label_int64.template Data<int64_t>();
+    for (size_t i = 0; i < (size_t)N; ++i)
+      label->template MutableData<std::string>()[i] = classlabels_strings_[plabel[i]];
```

Comment:
>label->template MutableData<std::string>()[ [](start = 6, length = 43)

Can we call MutableData once outside the loop? Not sure if the compiler can optimize away the ORT_ENFORCE that's in there. Would be interesting to check the disassembly from a release build.

Reply:
fixed